### PR TITLE
Oblivious attribution protocol [part 1 - attribution]

### PR DESCRIPTION
**Pull request by taikiy**
# Overview
This diff implements the first part (out of two - "attribution" and "aggregation") of the last touch oblivious attribution in a secure multiparty computation. Currently it supports "Last Touch" attribution only. Produced result will be used in the second part - "aggregation", which requires sorting of the list.

# Details
Last touch attribution algorithm is described in details here:
https://github.com/patcg-individual-drafts/ipa/blob/main/IPA-End-to-End.md#oblivious-last-touch-attribution

First part of the protocol accesses and accumulates data of each node in a sorted list, and produces an intermediate result that has accumulated credits for each breakdown key. The document also describes how the logic can be optimized as follows:
```
flag = current.stop_bit * successor.helper_bit * successor.trigger_bit
current.credit = current.credit + flag * successor.credit
current.stop_bit = flag * successor.stop_bit
```
This is what's implemented in `last_touch_attribution()` with multiplications done on a MPC. I believe there could be more optimizations if we put more constrains on the input, but for this round this basic/general implementation should be enough for us to measure baseline benchmarks.

### CODE REVIEW

**Comment by taikiy** [comment]
Rust check workflow failing because of a clippy error in a code comment. I'll fix this in the next commit when I address feedbacks.

File: src/protocol/attribute.rs
Comment (martinthomson) [inline src/protocol/attribute.rs:11]:
```suggestion
/// Oblivious attribution protocol
```

File: src/protocol/attribute.rs
Comment (martinthomson) [inline src/protocol/attribute.rs:81]: I find this ordering of the arrays surprising.  Why do we have an array of arrays where the outer array is indexed by the field type?  Why are we using array indices to access shares/fields anyway?

I had expected that we would be operating on a sequence of structs, where each field in the struct is a replicated share.  This looks like a sequence of sequences, where the outer sequence has some sort of implicit structure.  That leads to some odd potential failure modes that need to be guarded against, like where the Vec for the helper bit is a different length to the Vec for the stop bit.

File: src/protocol/attribute.rs
Comment (martinthomson) [inline src/protocol/attribute.rs:17]: "Attribute" is maybe not the best choice of name for this, given the potential for it to be confused.  "Attribution" is probably cleaner, even if it doesn't have the right tense.

File: src/protocol/attribute.rs
Comment (martinthomson) [inline src/protocol/attribute.rs:120]: Is this a function that we should be providing on the Gateway?

File: src/protocol/attribute.rs
Comment (martinthomson) [inline src/protocol/attribute.rs:134]: I think that you want to use zip() here instead:

```rust
a.zip(b).map(async move |(a, b)| self.ctx.multiply(RecordId::from(i), self.step).await.execute(a, b).await)
```

File: src/protocol/attribute.rs
Comment (martinthomson) [inline src/protocol/attribute.rs:19]: This isn't going to work, I don't think.  The step needs to be taken from the same global enum.  So in order to have this work, you need to take the top-level enum we're using.  Which means parameterizing on `S: Step` and taking an argument of that type.

I also think that we will probably want to make the implementation of attribution a generic parameter.  That is, you might define a generic attribution step as a trait, then a concrete implementation of that would be last touch.

**Review by martinthomson — CHANGES_REQUESTED** [review]
Just a short review, based on a few things that I noticed.

File: src/protocol/attribute.rs
Comment (benjaminsavage) [reply src/protocol/attribute.rs:19]: Yeah, I'd recommend just focusing on this one attribution algorithm in isolation for now. Once we have it running, we can think about how to support other stuff. I'd call this module "LastTouchAttribution"

File: src/protocol/attribute.rs
Comment (taikiy) [reply src/protocol/attribute.rs:81]: I was just discussing this with Richa and others. I agree that we expect to receive a sequence of structs. When it comes to the actual protocol executions, we realized we were always processing a sequence of replicated shares for a certain field (i.e. match key). For this PR, I chose this input format for the sake of simplicity, assuming the input had been transposed. Should all top-level protocol steps expect that the input format would be a sequence of structs?

File: src/protocol/attribute.rs
Comment (taikiy) [reply src/protocol/attribute.rs:120]: Makes sense. Many protocols seem to need this, so let's put it in the gateway.

File: src/protocol/attribute.rs
Comment (benjaminsavage) [reply src/protocol/attribute.rs:81]: I'm confused by your reply taiki.

I think Martin is suggesting something like:

struct AttributionInputs {
    helper_bit: Replicated<F>,
    is_trigger_bit: Replicated<F>,
    value: Replicated<F>,
    breakdown_key: Replicated<F>,
}

And the "execute" function would accept a Vec of these things.

File: src/protocol/attribute.rs
Comment (benjaminsavage) [inline src/protocol/attribute.rs:98]: OK, I see what you're doing here. It's very subtle. The index "i" refers to the correct successor here because the b_cred vector is shifted compared to the original input.

That is very clever, but I'm worried its a bit too subtle. It's harder to understand and I worry will lead to bugs.

Honestly, I would rather see something like:

b_cred = foo[i] + bar[i + step_size];

I feel like that makes it easier to see what's going on.

File: src/protocol/attribute.rs
Comment (benjaminsavage) [reply src/protocol/attribute.rs:120]: @richajaindce could have used this in GenBitPerm

File: src/protocol/attribute.rs
Comment (benjaminsavage) [inline src/protocol/attribute.rs:171]: Yeah, I have no idea what this means. I think the structs will help.

File: src/protocol/attribute.rs
Comment (taikiy) [reply src/protocol/attribute.rs:81]: The code was bit more readable to write

```
.multiply_all(
    &self.input[Self::_HELPER_BIT][step_size..],
    &self.input[Self::_IS_TRIGGER_BIT][step_size..],
)
```

than

```
impl Index for AttributionInputs { ... }

.multiply_all(
    &self.input[step_size..],
    SELF::_HELPER_BIT,
    SELF::_IS_TRIGGER_BIT,
)
```

that's all. I'm okay with the latter.

As for the sort protocol, I believe it creates a vector for each sort key. @richajaindce, your thoughts?

Comment (richajaindce) [reply src/protocol/attribute.rs:81]: It looks like we always operate on same type. This is true for running attribution and also if we want to execute a shuffle data attached to matchkey before sort (to hide information).  Would keeping all of data in memory be an issue? Using Taiki's approach, not everything needs to be in memory vs event approach needs all data to be in memory.

File: src/protocol/attribute.rs
Comment (akoshelev) [inline src/protocol/attribute.rs:136]: lets add `debug_assert` for that maybe?

File: src/protocol/attribute.rs
Comment (akoshelev) [inline src/protocol/attribute.rs:58]: I would much rather prefer if we turn this assumption into a type, i.e. use u32 here and below. In this case we don't need to keep being cautious about pieces of code that assume range fits into u32 but does not validate it. 

This unfortunately leads to major inconveniences working with vector indices and vector capacity as we have to use usizes for it. 

@martinthomson @benjaminsavage I wonder if we should create a wrapper around Vec and implement `Index<u32>/IndexMut<u32>` for it? We will enforce that this vector never grows over $2^{32}-1$ elements but working with it will be much more ergonomic and type safe than standard vec.

File: src/protocol/attribute.rs
Comment (akoshelev) [inline src/protocol/attribute.rs:75]: you can replace these three lines (init, while, *= 2) with a one liner that creates an iterator over powers of 2. 

```rust
        // generate powers of 2 that fit into input len. If num_rows is 15, this will produce [1, 2, 4, 8]
        for step_size in std::iter::successors(Some(1u32), |prev| prev.checked_mul(2)).take_while(|&v| v < num_rows) {
}
```

File: src/protocol/attribute.rs
Comment (martinthomson) [reply src/protocol/attribute.rs:58]: Yes, I think that a wrapper would do a lot to help.

Comment (martinthomson) [reply src/protocol/attribute.rs:58]: Oh, I should say: the index type should probably be a typedef, like `type RecordIndex = u32` or something along those lines.

File: src/protocol/attribute.rs
Comment (taikiy) [reply src/protocol/attribute.rs:81]: I am going to fix this part to take a slice of structs, but just wanted to clarify what's going on in the code.
So, the [attribution algorithm](https://github.com/patcg-individual-drafts/ipa/blob/main/IPA-End-to-End.md#oblivious-last-touch-attribution) would look like this:

```
// input: Vec<GenericReport>
let mut credit: Vec<Replicated<F>> = input.map(|x| x.value).collect();
let mut stop_bit: Vec<Replicated<F>> = [F::ONE; input.len()];

while step_size < input.len()
    let mut new_credit = credit.clone();
    let mut new_stop_bit = stop_bit.clone();

    for i in 0..input.len() {
        if i+step_size > input.len() break;

        let current = i;
        let successor = i+step_size;

        // in reality these will be lengthy async multiply calls
        let flag = stop_bit[i] * input[successor].helper_bit * input[successor].trigger_bit;
        new_credit[current] = credit[current] + flag * credit[successor];
        new_stop_bit[current] = flag * stop_bit[successor];
    }

    // `new_credit` is (input.len() - step_size) long, but `credit` must be input.len() length
    let end = input.len() - step_size;
    credit = new_credit[0..end].append(credit[end..]);
    // same for stop_bit
    stop_bit = new_stop_bit[0..end].append(stop_bit[end..]);
    
    step_size *= 2;
}
...
// we drop unnecessary fields. only need to return aggregated credits
Ok([{breakdown_key, credit}; unique(breakdown_key).len()])
```

Note that a `for` loop iteration doesn't depend on the previous iterations. Since they are independent we can eliminate the `for` loop and execute multiplications on slices of shares concurrently, which is what `multiply_all` does. It also eliminates awkward boundary checks, reassign of variables, padding vectors, etc.

File: src/protocol/attribute.rs
Comment (taikiy) [reply src/protocol/attribute.rs:171]: Yeah, this is my fault. I didn't add any comments to this and some arrays are irrelevant. I'll clarify in this thread. https://github.com/martinthomson/raw-ipa/pull/108#pullrequestreview-1136724098

File: src/protocol/attribution/accumulate.rs
Comment (taikiy) [inline src/protocol/attribution/accumulate.rs:17]: Renamed from `Attribution` to `Accumulate` to better reflect what it is actually doing. There are three steps in the oblivious attribution. 1. accumulation (this step), 2. sorting, 3. aggregation. I will create Attribution trait in future PR when I implement the rest of the steps.

File: src/protocol/attribution/accumulate.rs
Comment (martinthomson) [reply src/protocol/attribution/accumulate.rs:298]: Try `(0..=4).map(|i| AccumulateTestStep::Step(i))` if you want to type less.

File: src/protocol/attribution/accumulate.rs
Comment (benjaminsavage) [inline src/protocol/attribution/accumulate.rs:78]: There is a bit of a tricky thing with the rows that are near the end. They don't need any more accumulation, but you need to pass them along somehow.

e.g. if there are 31 rows, then on iteration 5 you're going to have step_size = 16, and end = 15. So there will be just 15 objects in the "accumulation_futures" vector. Those have accumulated credit from the first 4 iterations that needs to not get lost.

File: src/protocol/attribution/accumulate.rs
Comment (benjaminsavage) [reply src/protocol/attribution/accumulate.rs:17]: Maybe AccumulateCredit? Since "accumulate" is kind of a general term.

File: src/protocol/attribution/accumulate.rs
Comment (taikiy) [reply src/protocol/attribution/accumulate.rs:78]: Following your example, do you mean at step_size = 16, end = 15 and accumulating credit at row i = 0, I need to pass credit[16] so that I can calculate `credit[0] + b * credit[16]`?

There is a separate `credits` vector that keeps track of the accumulated credits (line 110). When I create `current` and `successor` structs, I copy values from this `credits` (line 84 and 89) which have accumulated credits from the previous iteration.

Comment (benjaminsavage) [reply src/protocol/attribution/accumulate.rs:78]: Oh, I see. You're absolutely correct. The "credits" mutable vector should have everything in it.

File: src/protocol/attribution/accumulate.rs
Comment (benjaminsavage) [inline src/protocol/attribution/accumulate.rs:139]: Rather than pass in the step size (which it doesn't really need to know), perhaps just pass in a boolean indicating if this is the first step or not.

File: src/protocol/attribution/accumulate.rs
Comment (benjaminsavage) [inline src/protocol/attribution/accumulate.rs:99]: This is really not awesome. Fortunately this is probably all going away as a result of @martinthomson's latest PR.

File: src/protocol/attribution/accumulate.rs
Comment (taikiy) [reply src/protocol/attribution/accumulate.rs:139]: Yes, I was wondering what would be the most non-absurd way to pass this. I thought `if step_size > 1` was a tiny bit more readable than `if !first_iteration`, but I'm okay with a boolean, too.

File: src/protocol/attribution/accumulate.rs
Comment (benjaminsavage) [inline src/protocol/attribution/accumulate.rs:105]: I get thrown off by the term "support vectors" since it makes me think of https://en.wikipedia.org/wiki/Support_vector_machine

I'd suggest:

// accumulate the credit from this iteration into the accumulation vectors

File: src/protocol/attribution/accumulate.rs
Comment (taikiy) [reply src/protocol/attribution/accumulate.rs:99]: Agree 100%. I'll probably wait for Martin's PR to merge and improve this part. I think I actually need different steps for each iteration (step_size). I do `try_join_all` before I go to the next iteration but that's not sync'ed across the helpers, so things might get tangled up.

File: src/protocol/attribution/accumulate.rs
Comment (benjaminsavage) [reply src/protocol/attribution/accumulate.rs:139]: You could do something similar to what @richajaindce did here: https://github.com/martinthomson/raw-ipa/pull/97/files#diff-894d6d54e26cbe3155f95c21f6bd6b557778f2705f8371fc2bc1d31b2292cb74

She made functions accept an enum - which only had two values. Direction::LEFT and Direction::RIGHT.

This makes reading the call site easier to understand than seeing the function accept "true" or "false" for the nth argument. You could have FooBar::FIRST_ITERATION and FooBar::SUBSEQUENT_ITERATION or something.

File: src/protocol/attribution/accumulate.rs
Comment (benjaminsavage) [inline src/protocol/attribution/accumulate.rs:172]: What does this do?

File: src/protocol/attribution/accumulate.rs
Comment (taikiy) [reply src/protocol/attribution/accumulate.rs:139]: Hmm, that might be an overkill for this case. Rust doesn't support ternary conditional operator so the call site would then look like this.

```
let iteration = if step_size > 1 { FooBar::FIRST_ITERATION } else { FooBar::SUBSEQUENT_ITERATION }; // need the brackets, too

get_accumulated_credit(
    ctx,
    iteration,
    ...
```

Comment (benjaminsavage) [reply src/protocol/attribution/accumulate.rs:139]: You could have a mutable local variable initialized to FooBar::FIRST_ITERATION, which is updated to FooBar::SUBSEQUENT_ITERATION at the end of the loop. Then you don't need a ternary.

File: src/protocol/attribution/accumulate.rs
Comment (taikiy) [reply src/protocol/attribution/accumulate.rs:172]: `Either::Left` and `Either::Right` let you return futures conditionally. For this case, I want to skip the multiplication in the first iteration, so I return a future that resolves to `b` immediately when awaited.

File: src/protocol/attribution/accumulate.rs
Comment (taikiy) [reply src/protocol/attribution/accumulate.rs:139]: Or maybe I could use the `steps` to infer which iteration I'm in? As I mentioned in a separate comment, I think I need 4 * log(n) steps rather than just 4. I need to look into Martin's PR to see if that's feasible.

**Review by benjaminsavage — APPROVED** [review]
This looks good to me. I recommend you coordinate with @martinthomson to see if you should land this first, or wait until he lands his first.

File: src/protocol/batch.rs
Comment (taikiy) [inline src/protocol/batch.rs:7]: This wrapper ensures that the length of the inner vector is less than RecordIndex::MAX upon index references, but doesn't prevent it from growing larger at the time items are added (e.g., push(), append()). Is there a way to monitor changes in its size? @martinthomson @akoshelev

Comment (martinthomson) [reply src/protocol/batch.rs:7]: It might be better to build this without exposing the internal collection, so that you only need to cover a few different interfaces for manipulation.

I imagine that you want to support `Iterator::collect()`, but maybe you don't need push, extend_from_slice() and so forth.

**Comment by taikiy** [comment]
Rebase onto @martinthomson's renewed Step scheme. The most recent commit to be reviewed: https://github.com/martinthomson/raw-ipa/pull/108/commits/9e4d5378e13bd9424340876804e8dbe1b0603580

File: src/protocol/attribution/accumulate_credit.rs
Comment (akoshelev) [inline src/protocol/attribution/accumulate_credit.rs:102]: @martinthomson in many protocols there seem to be a pattern of generating large number of futures and then polling them all. Do you see it as a problem, given that no work is done until the first poll?

File: src/protocol/attribution/mod.rs
Comment (akoshelev) [inline src/protocol/attribution/mod.rs:45]: there will be $O(N)$ allocations here, we probably want to mark this as todo to fix. I think @martinthomson was planning to add a cheap way to create a unique step id for release builds

File: src/protocol/attribution/accumulate_credit.rs
Comment (akoshelev) [inline src/protocol/attribution/accumulate_credit.rs:18]: this input format implies that somebody after doing the sort must convert the regular batch of sorted secret shares into this "expanded" format with helper bits. Is the plan to insert an "intermediate" step between sort and attribution to convert shares format or you want `AccumulateCredit` to eventually implement the conversion?

File: src/protocol/attribution/accumulate_credit.rs
Comment (akoshelev) [inline src/protocol/attribution/accumulate_credit.rs:69]: Please correct me if I am wrong, but it seems we are using `record_id` as step and `iteration_step` as record_id. Is it intended?

File: src/protocol/attribution/accumulate_credit.rs
Comment (akoshelev) [inline src/protocol/attribution/accumulate_credit.rs:45]: I wonder if it is possible to make credit and stop bit part of the input?

File: src/protocol/attribution/accumulate_credit.rs
Comment (taikiy) [reply src/protocol/attribution/accumulate_credit.rs:69]: Yes, that is correct. This is the issue @benjaminsavage and I were talking about during the meeting. Since this is a sub-protocol that executes 4 multiplications on a single row, I wanted to pass a reference to the ctx, set 4 multiplication steps like Step::Mult1, Step::Mult2, ..., and each row (parallelized) have different record IDs (i.e. row i = RecordId::from(i)). With this set up, in the sub-protocol, the step should look like `protocol/iteration_1/mult_1` and `recrod_id=i` where i in `(0..input.len())`. But when I try to run that, I get this error:

```
thread 'protocol::attribution::accumulate_credit::tests::accumulate' panicked at 'Refined 'protocol' with step 'mult_1' twice', src/protocol/mod.rs:94:13
```

So, I ended up with this diff, which calls another `ctx.narrow()` for each row to give them unique steps (i.e. "row_i"), and each multiplication as separate records.

File: src/protocol/attribution/mod.rs
Comment (taikiy) [reply src/protocol/attribution/mod.rs:45]: This should go away once we decide to move away from `AsRef<str>`. I understand AsRef<str> for steps is temporary for this prototype?

File: src/protocol/attribution/accumulate_credit.rs
Comment (taikiy) [reply src/protocol/attribution/accumulate_credit.rs:18]: Helper bits will be a part of the sort output. IIUC, the sort protocol will output a vector or vectors of index which corresponds to the sorted locations of the input list. I think there will be an intermediate step where we might remove unnecessary fields like match keys and timestamps, but that is still unknown.

File: src/protocol/attribution/accumulate_credit.rs
Comment (taikiy) [reply src/protocol/attribution/accumulate_credit.rs:45]: I can see them being part of the input if we have a parent protocol that calls `AccumulateCredit` after `Sort`, but we don't have that yet. Credits and stop bits are the states/results generated by running this protocol, so it seems reasonable to create them inside this protocol.

File: src/protocol/attribution/accumulate_credit.rs
Comment (benjaminsavage) [reply src/protocol/attribution/accumulate_credit.rs:18]: I have this strong suspicion that generating the "helper bits" *within* the MPC will be more efficient than trying to generate them in a separate step.

File: src/protocol/attribution/accumulate_credit.rs
Comment (benjaminsavage) [reply src/protocol/attribution/accumulate_credit.rs:45]: I agree with @taikiy - these are basically local variables to this protocol. They're all initialized to Replicated::one(), or one of the other inputs at the beginning of the loop (and over-written in each iteration) so there's no value in passing them in.

File: src/protocol/attribution/accumulate_credit.rs
Comment (benjaminsavage) [reply src/protocol/attribution/accumulate_credit.rs:69]: We shouldn't have to do this. This is an issue with where that "check" happens in the code. Looking forward to an infra fix that helps avoid the need to do this!

File: src/protocol/attribution/accumulate_credit.rs
Comment (benjaminsavage) [inline src/protocol/attribution/accumulate_credit.rs:148]: This seems wrong. This should be a proper record_id, not just 0 or 1. This is related to the same infra issue from above - we need the ability to run N multiplications in parallel, with the same step, and only the record_id differing.

**Review by benjaminsavage — CHANGES_REQUESTED** [review]
This is all looking really good, with the exception of needing to work around that infra issue that prevents us from using the same step name on different records.

**Pull request merged**
